### ISSUE TITLE
Move longTasksMetric out of system_health/

### ISSUE THREAD

**Issue by benshayden** [catapult-project/catapult#3158]
This metric isn't used by any benchmarks, but it is a great demo of the APIs and UI.
I think it's also a useful trace viewer feature to have all of the long tasks measured and collected so you can drill down into them.

@nedn @eakuefner

**Comment by nedn** [comment]
How is this related to expected input latency metric? Maybe diagnostic metric?
@tdresser

**Comment by tdresser** [comment]
If anything, this is a diagnostic of Expected Queueing Time. 

I agree that this is a useful trace viewer feature. Once we've shipped the long tasks web perf API, developers will need to be able to quickly find all their long tasks, and we may want something like this built into devtools.

**Comment by benshayden** [comment]
If this will be a diagnostic of EQT (which is a SH metric IIUC), then should it stay in system_health/?
If so, then you can close this. :-)

**Comment by nedn** [comment]
Maybe retitle this to "modifying long task metric to be diagnostic of EQT"?

**Comment by benshayden** [comment]
I think it would make sense to keep longTasksMetric as a top-level metric as well as adding it as a diagnostic of EQT, so that it can remain a top-level feature of trace viewer. Any objection?

Does @ulan or @deepanjanroy own EQT?

**Comment by tdresser** [comment]
What exactly do you mean by top level metric?

I think that probably means different things in different contexts.

By top level metric in this case, you mean "easily visible in tracing"?

That sounds reasonable to me, along with adding it as an EQT diagnostic.

**Comment by benshayden** [comment]
Yes, sorry, I meant "registered metric", that isn't necessarily a source node of the metric graphical model depending on how the benchmark is configured, but is accessible directly in the metrics-side-panel.

**Comment by tdresser** [comment]
SGTM.

In terms of ownership, I expect dproy@ will likely end up owning EQT in general in the long term (if that's okay with ulan@), but ulan@ may have more ownership of it currently, as he did a bunch of work getting it working for v8 folks.

**Comment by ulan** [comment]
Ben, can you please find another owner for this issue (if it is still relevant)? I don't have free cycles currently.

**Comment by ulan** [comment]
> In terms of ownership, I expect dproy@ will likely end up owning EQT in general in the long term (if that's okay with ulan@)
That would be definitely ok with me :)

**Comment by tdresser** [comment]
@maxlgu has been driving C++ side EQT, assigning to him (though it's fairly low priority).

**Comment by maxlgu** [comment]
But currently each metric is independently calculated. I think we don't want duplicate calculation of longTask in EQT.

Is it a OK that we fetch EQT histogram and add longtask histograms to it as a diagnostic, after we finish all the calculations of metrics? I concern about whether it's a good structure, seems each 'metric'.html is supposed to know all its values, diagnostics and histogram in itself.  Or do we have other better ideas?

**Comment by benshayden** [comment]
Can you rely on benchmarks adding metrics in the right order? I.e. ensure that CreateTimelineBasedMeasurementOptions adds longTasksMetric before EQT metric?
If so, then EQT can look for longTasks histograms in the HistogramSet and add the diagnostics if it finds them.

If that would be too complicated, we have ideas for other solutions depending on how else you need this to work. In particular, should longTasksMetric always be run whenever EQT is run?

**Comment by maxlgu** [comment]
EQT and longTasksMetric does not always run together. Metrics run by trace viewer only execute one metric at a time.

**Comment by maxlgu** [comment]
Also, are these metrics used by other things than trace viewer and benchmarks?

**Comment by maxlgu** [comment]
Discussed offline with benshayden@, we have several approaches available:
1. Assuming that EQT depends on long task for diagnostics, if we are computing EQT and find that long task is not in the histogram set, we add long task in the histograms and add it to diagnostic. Otherwise we just add it to diagnostic without creating long task histogram. 
2. We can define "meta metric", i.e. 'EQTLongtask matric' to add EQT, long task and add long task as a diagnostic for EQT.
3. We can define dependencies between each pair of dependent metrics.

**Comment by tdresser** [comment]
Do we expect dependencies to be common?

If we take option 1, presumably if EQT is computed first, and long tasks second, then long tasks would just return early, as EQT will have already populated long tasks?

If so, option 1 SGTM.

**Comment by maxlgu** [comment]
In option 1 we assume there is a order, if EQT depends on long task, then long task should be executed first, vice versa. But what if in the future we also need EQT to be a diagnostic in long task? Then EQT and long task both assume each other executed first. So option 1 cannot solve this future need.

**Comment by maxlgu** [comment]
> If we take option 1, presumably if EQT is computed first, and long tasks second, then long tasks would just return early, as EQT will have already populated long tasks?

Discussed offline with tdresser@, this have two problems:
1. it indicates that a metric can execute another metric within itself, which doesn't seem like a good structure.
2. in metric-side-panel, only one metric is executed at a time. We need long task as a dependency of EQT but we don't need to add long task into histogram set. 

Can we compute all the metrics as well as their dependencies first, and then decides which goes to histograms and which goes to metrics?

**Comment by benshayden** [comment]
1. It's probably ok to allow metrics to call other metrics. If we find it causing problems, then we can address those problems, possibly using a dependency graph, possibly something else simpler. We've had metametrics before, and only removed them because we decided it was simpler to allow benchmarks to ask TBM to run multiple metrics.
2. The long tasks histograms *do* need to be added to the same HistogramSet as EQT's histograms: metrics-side-panel can only pass a single HistogramSet to histogram-set-view.
3. Running all metrics could take a few seconds, which seems like a waste if their results will be inaccessible, just in case there's a hidden dependency. Explicit is better than implicit. It would be technically feasible for metrics-side-panel to keep track of which metrics produce which histograms, but it seems more complex than explicitly annotating dependencies. Dependency graphs are a bit complex, but it would be far from the most complex thing about TBMv2 -- that will probably always be merging histograms and diagnostics, especially relationships.
